enhancements/testing/openshift-tests-extension.md

```diff
@@ -0,0 +1,1202 @@
+is non-trivial to implement, so to simplify the creation of extensions, a Go module &
+repository will be maintained by TRT: https://github.com/openshift-eng/openshift-tests-extension .
+
+Vendoring this module and following its integration guide will provide component test authors
+the majority of the logic necessary to fulfill the test extension interface. It will
+also provide a means of centrally defining new testing requirements. As component owners re-vendor
+the module, they would be required to implement those new testing requirements
+for the module to compile and run a test extension binary successfully.
+
+During a run of the origin test framework (`openshift-tests`), the framework will:
+
+1. Discover available extension binaries.
+2. Discover metadata and component configuration options for those extensions (running `info` verb and parsing standard
+   out).
+3. Discover tests available from those extensions (running `list` verb and parsing standard out).
+4. Plan an appropriate and efficient testing strategy based on the framework invocation.
+5. Execute the tests using subsequent invocations of the discovered binaries (`run-test` invocations).
+6. Collect the result and output of the test invocations and integrate it into overall test suite results.
+
+#### Binary Discovery
+
+The existence of test extension binaries can be registered one of two ways by test authors.
+
+##### OpenShift Payload Extension Binaries
+
+For OpenShift payload components contributors can advertise the existence of an extension binary
+by adding information (the imagestream tag for the OCP payload component and the path to the binary
+within their image) to a simple registry datastructure in github.com/openshift/origin.
+
+##### Non-Payload Extension Binaries
+
+For non-payload components, contributors must advertise their extension binary for `openshift-tests`
+to discover. This is accomplished by creating an ImageStream / ImageStreamTag with a special label:
+
+`testextension.redhat.io/component=<product-name>-<component-name>`
+
+An annotation on the ImageStreamTag will then identify the binary to extract and (optional)
+arguments to pass ahead of the extension interface verbs.
+`testextension.redhat.io/binary=<binary-path>.gz [--argument=value]`
+
+Well-behaved operators should gate the creation of this imagestream/label/annotation on the existence of the
+`TestExtensionAdmission` custom resource definition (the code should not attempt to
+inspect `TestExtensionAdmissions` instances -- just check for the CRD itself so that operators
+don't liter clusters with ImageStreams only applicable for testing in a production environment).
+
+Optional Operator authors must ensure that the image carrying the extension binary
+is identified in their ClusterServiceVersion (CSV) so that tools like `oc-mirror`
+will copy image(s) bearing extension binaries to disconnected clusters.

```

Comment:
(starting a separate thread). To get the name of a mirrored image, a test is suggested to call `github.com/openshift/origin/test/extended/util/image.LocationFor("my.source/image/location:versioned_tag")` [here](https://github.com/openshift/origin/blob/b573a5bc983db72afd094614f5f61f725b975948/test/extended/util/image/README.md?plain=1#L20). Does the extension need to import `openshift/origin`? That could be problematic, especially if the extension imports incompatible version of `k8s.io/kubernetes`.

Reply:
This is something we'll need to solve for sure.  We want to minimize vendoring requirements in the component repos for sure.  I do not want anyone to have to vendor origin code.

Perhaps we could pass the image locations to each tests binary and have OTE provide a LocationFor helper.  It could be passed either as a CLI flag or a path to a JSON/YAML file.